File: leetcode/leetcode-23.py
from typing import List
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        import heapq
        k = len(lists)
        head = []
        for i in range(k):
            if lists[i]:
                heapq.heappush(head, (lists[i].val, i))
                lists[i] = lists[i].next
        ans = ListNode(0)
        que = ans
        while head:
            v , i = heapq.heappop(head)
            que.next = ListNode(v)
            que = que.next
            if lists[i]:
                heapq.heappush(head, (lists[i].val, i))
                lists[i] = lists[i].next
        return ans.next
        # flag = True
        # u = -1
        # v = 1000000000
        # for i in range(k):
        #     if lists[i] is not None:
        #         if v > lists[i].val:
        #             u = i
        #             v = lists[i].val
        # if u != -1:
        #     ans = ListNode(v)
        #     lists[u] = lists[u].next
        # else:
        #     return None
        # que = ans
        # while flag:
        #     u = -1
        #     v = 1000000000
        #     for i in range(k):
        #         if lists[i] is not None:
        #             if v > lists[i].val:
        #                 u = i
        #                 v = lists[i].val
        #     if u == -1:
        #         break
        #     t = ListNode(v)
        #     lists[u] = lists[u].next
        #     que.next = t
        #     que = que.next
        # return ans

if __name__ == "__main__":
    q = [[1,4,5],[1,3,4],[2,6]]
    so = Solution()
    li = []
    a = ListNode(1)
    b = ListNode(4)
    c = ListNode(5)
    a.next = b
    b.next = c
    li.append(a)
    a = ListNode(1)
    b = ListNode(3)
    c = ListNode(4)
    a.next = b
    b.next = c
    li.append(a)
    a = ListNode(2)
    b = ListNode(6)
    a.next = b
    li.append(a)
    ans = so.mergeKLists(li)
    while ans:
        print(ans.val, end=' ')
        ans = ans.next

File: leetcode/leetcode-23.cpp
/*
 *@File Name          :leetcode-23.cpp
 *@Author             :cc
 *@Version            :1.0.0
 *@Date               :2022/2/25 9:07
 *@Description        :
 *@Function List      :
 *@History            :
 */

#include "algorithm"
#include "string"
#include "vector"

using namespace std;


struct ListNode {
    int val;
    ListNode *next;

    ListNode() : val(0), next(nullptr) {}

    ListNode(int x) : val(x), next(nullptr) {}

    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


class Solution {
public:
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        return dfs(0, lists.size() - 1, lists);
    }

    ListNode *dfs(int l, int r, vector<ListNode *> &lists) {
        if (l > r) return nullptr;
        if (r == l) return lists[l];
        int mid = (r + l) / 2;
        ListNode *t1 = dfs(l, mid, lists);
        ListNode *t2 = dfs(mid + 1, r, lists);
        ListNode *ans = new ListNode();
        ListNode *anss = ans;
        while (t1 && t2) {
            if (t1->val < t2->val) {
                ans->next = t1;
                t1 = t1->next;
            } else {
                ans->next = t2;
                t2 = t2->next;
            }
            ans = ans->next;
        }
        if (t1) {
            ans->next = t1;
        }
        if (t2) {
            ans->next = t2;
        }
        return anss->next;
    }
};